nengo_ocl/clra_gemv.py

```diff
@@ -1316,16 +1662,380 @@ def plan_sparse_dot_inc(queue, A_indices, A_indptr, A_data, X, Y, inc=False, tag
+
+    Returns
+    -------
+    Plan
+        Executable Plan object
+    """
+    kern = """
+    __kernel void ellpack_inc(
+        __global const ${IndType} *A_columns,
+        __global const ${dtype} *A_entries,
+        __global const int *Xstarts,
+        __global const ${dtype} *Xdata,
+        __global const int *Ystarts,
+        __global ${dtype} *Ydata
     )
+    {
+        // n can later be used to keep track of multiple arrays
+        const int n = 0;
+        const int irow = get_group_id(0);
+        const int iellcol = get_local_id(1);
+
+        __global const ${dtype} *x = Xdata + Xstarts[n];
+        __global ${dtype} *y = Ydata + Ystarts[n];
+
+        __local ${dtype} products[${wg_size}];
+
+        // Load into individual products, doing multiplication along the way
+        if (iellcol < ${ellwidth}) {
+            const ${dtype} weight = A_entries[irow * ${ellwidth} + iellcol];
+            const ${IndType} jdatacol = A_columns[irow * ${ellwidth} + iellcol];
+            products[iellcol] =  weight * x[jdatacol];
+        } else {
+            products[iellcol] = 0;
+        }
+        barrier(CLK_LOCAL_MEM_FENCE);
+
+        // Do reduction across work group
+    %if serial_reduction:
+        if (iellcol == 0) {
+            for (int k = 1; k < ${ellwidth}; k++) {
+                products[0] += products[k];
+            }
+        }
+    %else:
+        for (int offset = ${wg_size}/2; offset > 0; offset >>= 1) {
+            barrier(CLK_LOCAL_MEM_FENCE);
+            if (iellcol < offset && iellcol + offset < ${wg_size}) {
+                products[iellcol] += products[iellcol + offset];
+            }
+        }
+        barrier(CLK_LOCAL_MEM_FENCE);
+    %endif
+
+        // Store value
+        if (iellcol == 0) {
+    %if inc:
+            y[irow] += products[0];
+    %else:
+            y[irow] = products[0];
+    %endif
+        }
+    }
+    """
+    # --- dimensioning
+    nneurons, A_ellwidth = A_columns.shape
+    wg_size = round_up_power_of_2(A_ellwidth)
+    gsize = (nneurons, wg_size)
+    lsize = (1, wg_size)
+
+    # --- make the program
+    textconf = dict(
+        dtype=A_entries.ctype,
+        IndType=A_columns.ctype,
+        inc=inc,
+        ellwidth=A_ellwidth,
+        wg_size=wg_size,
+        serial_reduction=serial_reduction,
+    )
+    text = as_ascii(Template(kern, output_encoding="ascii").render(**textconf))
+    full_args = (
+        A_columns.base_data,
+        A_entries.base_data,
+        X.cl_starts.data,
+        X.cl_buf.data,
+        Y.cl_starts.data,
+        Y.cl_buf.data,
+    )
+    _fn = cl.Program(queue.context, text).build().ellpack_inc
+    _fn.set_args(*full_args)
+
+    plan = Plan(queue, _fn, gsize, lsize=lsize, name="cl_ellpack", tag=tag)
+    plan.full_args = full_args  # prevent garbage-collection
+
+    return plan
+
+
+def spmv_ellpacknonlocal_impl(queue, A_columns, A_entries, X, Y, inc=False, tag=None):
```

Comment:
At some point, we'll want to remove the non-performant ones. What do you think about removing them now, but doing it as a second commit, so that anybody wanting to see the other alternatives can still look at your original commit?